Docker Console and Debugger (#31)
navapbc/template-application-rails: docs/app-rails/technical-foundation.md

- Run the debugger:
  - Add a `debugger` line and the rails server will pause and start the debugger
  - If you're running the app natively, such as with `make start-native`:
    - You must connect to the debugger from another terminal session because of our [Procfile.dev](app-rails/Procfile.dev) configuration. 
    - From another terminal tab, run `rdbg -An`. 
    - If you have multiple Rails applications with debuggers running, you'll have to specify the port to attach the debugger to. For more information, see the [Rails debug gem documentation](https://github.com/ruby/debug?tab=readme-ov-file#remote-debugging).
  - If you're running the app in a container, such as with `make start-container`:
    - `rdbg` in the terminal on your host machine will not be able to see the port in the container to connect to the debugger. 

>     - Instead, run `rdbg` inside the container: `docker compose exec app-rails rdbg -An`

Reviewer: Running `rdbg -An` gave me this:

```
DEBUGGER (client): Connected. PID:10, $0:bin/rails
DEBUGGER (client): Type `Ctrl-C` to enter the debug console.
```

And I needed to hit `Ctrl-C` to enter the debug console. Whereas `rdbg -A` allowed me to skip hitting `Ctrl-C`. Any reason we need `-n`? I didn't see anything in particular in the docs.

Reviewer: Also, running `rdbg` in the container, I got this:

```
DEBUGGER (client): Connected. PID:10, $0:bin/rails

# No sourcefile available for /usr/local/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/single.rb
=>#0	[C] Thread#join at /usr/local/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/single.rb:63
  #1	Puma::Single#run at /usr/local/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/single.rb:63
  # and 20 frames (use `bt' command for all frames)

Stop by SIGURG
```

[This issue](https://github.com/puma/puma-dev/issues/309) is the only reference I found about this. I was able to hit `c` to continue past it, but it was persistent. Did you run into this?

Author: Thanks! i removed the `n` from the guidance, however that enables the --nonstop flag which is used to tell `rdbg` to not stop at the beginning of the script. I saw a bunch of troubleshooting tips that advised adding it, but it's not necessary in all cases.

I was only able to reproduce what you were seeing by connecting the debugger before the application hit the debug statement. But once the debug statement was hit by the user action, then the debugger terminal updated properly.

Reviewer: Oh I see! Thanks.

> 

Reviewer: Instead of:

```md
Rails has some useful [built-in debugging tools](https://guides.rubyonrails.org/debugging_rails_applications.html).
```

Replace with:

```md
Rails has some useful [built-in debugging tools](https://guides.rubyonrails.org/debugging_rails_applications.html). Here are a few different options:
```

> - Start the rails console: `make rails-console`

Reviewer: ```suggestion
- Start the [rails console](https://guides.rubyonrails.org/command_line.html#bin-rails-console): `make rails-console`
```

- Run a console in the browser:

>   - Add a `console` with `<% console %>` line to an `.erb` file and an interactive console, similar to the rails console, will appear in the bottom half of your browser window

Reviewer: ```suggestion
  - Add `<% console %>` to an `.erb` file and an interactive console, similar to the rails console, will appear in the bottom half of your browser window.
```

  - Note: If the console doesn't appear when running in a docker container, check to see if your IP address is added to the permissions list in [development.rb](app-rails/config/environments/development.rb) in `config.web_console.permissions`. The list is currently set to allow most internal IPs. You would also see an error in your terminal that looks something like: `Cannot render console from <your.IP.address.here>! Allowed networks: 127.0.0.0/127.255.255.255, ::1`
- Run the debugger:
  - Add a `debugger` line and the rails server will pause and start the debugger

>   - Note: If you're running the app with `make start-native` you'll have to connect to the debugger from another winder because of our [Procfile.dev](app-rails/Procfile.dev) configuration. From another terminal tab, run `rdbg -An`. If you have multiple Rails applications with debuggers running, you'll have to specify the port to attach the debugger to. Check out [Rails debug gem docks](https://github.com/ruby/debug?tab=readme-ov-file#remote-debugging) for more info.

Reviewer: ```suggestion
  - If you're running the app natively, such as with `make start-native`:
    - You must connect to the debugger from another terminal session because of our [Procfile.dev](app-rails/Procfile.dev) configuration. 
    - From another terminal tab, run `rdbg -An`. 
    - If you have multiple Rails applications with debuggers running, you'll have to specify the port to attach the debugger to. For more information, see the [Rails debug gem documentation](https://github.com/ruby/debug?tab=readme-ov-file#remote-debugging).
```

>   - Note: If you're running the app in a docker container with `make start-container` `rdbg` from your terminal will not be able to see the port in the container to connect to the debugger. You'll have to `exec` into the docker container and run the command from there. First, run `docker ps` and grab the `CONTAINER ID` for the container running your rails app. Then run this command to show the debugger: `docker exec -i <CONTAINER ID> rdbg -An`.

Reviewer: ```suggestion
  - If you're running the app in a container, such as with `make start-container`:
    - `rdbg` in the terminal on your host machine will not be able to see the port in the container to connect to the debugger. 
    - Instead, run `rdbg` inside the container: `docker compose exec app-rails rdbg -An`
